docs/design/Chem Integration.md

```diff
@@ -0,0 +1,66 @@
+# Chem integration into Upgrade Assistant
+
+Chem is a tool developed internally to solve the problem of matching loose assemblies (ie `.dll` files added into projects with unknown provenance). As customers are looking to move to .NET 5, this may pose a challenge as these versions are highly likely to be .NET Framework compilations that may not work as expected on newer platforms. Often, these libraries have NuGet packages now and newer versions that could work on newer platforms.
+
+The main idea behind Chem is based on the idea of chemical fingerprints (hence the naming) that some small amount of data may be used to infer the original identity. In the case of binaries, there are a number of fingerprints that may be useful in identifying what it may be:
+
+- File name
+- Public key token
+- File contents (and hashes thereof)
+- and others
+
+Work has already been done to generate a database of known packages along these and other identity dimensions. This data can then be used to generate potential indexes that could be used from within Upgrade Assistant to identify along this potential dimensions, both for notification and even to update a project file to reference the NuGet package rather than the loose assembly.
+
+## User Experience
+
+The main scenario this will focus on is one where a project has downloaded an assembly and added it to its repo and referenced it from various projects. The end result of this work should be able to:
+
+- Identify what package the dependency came from
+- Update a project file to now reference this from
+
+The expected UX for this will be to implement it as an extension that a user can opt into. As a user, in order to run the Chem integration, they will obtain the extension (ie `UpgradeAssitantChem.zip`) and then run as they normally would:
+
+```
+upgrade-assistant some-project.vbproj --extension .\path\to\UpgradeAssistantChem.zip
+```
+
+### Decoupling the data
+Initially, any data needed will be provided with the extension. This data would be what is used to map the fingerprints to actual packages. They often range around the 100mb-200mb range for useful data, but could be potentially larger if desired.
+
+However, it may be nice to decouple the data from the actual extension itself so they can be updated independently. In order to support his, Upgrade Assistant should provide a way for steps/extensions to provide step-dependent data. This could be something that a step could register it as having, and then Upgrade Assistant could provide commands such as the following:
```

Comment:
```suggestion
However, it may be nice to decouple the data from the actual extension itself so they can be updated independently. In order to support this, Upgrade Assistant should provide a way for steps/extensions to provide step-dependent data. This could be something that a step could register it as having, and then Upgrade Assistant could provide commands such as the following:
```